feat(cli): add YAML/ENV support to `airflow variables import`

evgenii-prusov opened the pull request:
# Feature Proposal: Add YAML Support to Airflow Variables Import Command

## Problem Statement

Apache Airflow currently has an inconsistency in its CLI commands for importing variables and connections:

- **`airflow connections import`**: Supports multiple file formats (JSON, YAML, ENV) using a flexible secret parser
- **`airflow variables import`**: Only supports JSON format with hardcoded `json.load()` and `json.dump()`

This inconsistency damages user experience and limits flexibility when managing Airflow variables, especially for teams that prefer YAML for configuration management.


This change brings the `airflow variables import` command to feature parity with `airflow connections import` by adding support for YAML and ENV file formats.

Key changes:
- Replace hardcoded `json.load()` with flexible `load_variables()` function
- Add format detection based on file extension (.json, .yaml, .yml, .env)
- Handle ENV files differently from JSON/YAML (single values vs. preserving data types)
- Add comprehensive tests for all supported formats

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

potiuk reviewed (approved):
LGTM - @Lee-W ?

shahar1 commented on airflow-core/src/airflow/secrets/local_filesystem.py:
```diff
@@ -233,10 +233,23 @@ def load_variables(file_path: str) -> dict[str, str]:
-    variables = {key: values[0] if isinstance(values, list) else values for key, values in secrets.items()}
+
+    # Determine a file format to handle lists correctly
+    ext = file_path.rsplit(".", 2)[-1].lower()
```
I'd rather use `Pathlib` method to determine file's suffix rather than string manipulation

Lee-W replied:
Same

shahar1 reviewed (commented):
Nice feature!
Resolving a small comment + adding some docs will do :)

Lee-W commented on airflow-core/src/airflow/cli/commands/variable_command.py:
```diff
@@ -81,20 +81,25 @@ def variables_import(args, session):
+    try:
+        # Use load_variables which supports JSON, YAML, and ENV formats
+        var_dict = load_variables(args.file)
+    except Exception as e:
```
Is it possible to catch a narrower exception?

Lee-W replied:
I think this is not resolved. What I meant is whether we could change `except Exception` to something like `except JSONDecodeError` but contains all the possible exceptions expected

potiuk replied:
I guess that one still needs to be resolved?

Lee-W replied:
yep, I think this is still missed

evgenii-prusov replied:
I've reimplemented the proposed change. However, during development I discovered an important limitation:
The current load_variables() implementation in LocalFilesystemBackend doesn't support the description property for variables. This means that moving the CLI import to use this shared logic removes the existing ability to import variables with descriptions from **JSON files**.

**Decision Needed:**
We need to decide the priority between:
1. New capability: YAML and ENV file support for variable imports
2. Existing capability: JSON files with variable descriptions

**Options:**
- Accept the breaking change and document it
- Revert to custom JSON parsing in CLI while keeping error handling improvements
- Enhance LocalFilesystemBackend.load_variables() to support descriptions

Please advise on the preferred approach. The current implementation is ready but introduces this breaking change.

P.S. @Lee-W, parsing level exceptions `yaml.MarkedYAMLError`, `JSONDecodeError` handled in parser functions `_parse_yaml_file` and `_parse_json_file` respectively and transformed into `AirflowFileParseException` later. As for FileNotFound and other OSErrors they are handled directly in `variables_import` function.

evgenii-prusov replied:
I also found this old https://github.com/apache/airflow/issues/37532 which shows that working with variable descriptions had limitations even before this change. The issue highlighted that variable exports from the UI didn't include description fields, requiring manual updates when importing to other environments. This suggests the description functionality was already incomplete in the broader ecosystem.

Happy to tackle the description functionality improvements if you're interested and can point me in the right direction.

Lee-W replied:
>    Enhance LocalFilesystemBackend.load_variables() to support descriptions
 

ideally this

Lee-W replied:
> parsing level exceptions yaml.MarkedYAMLError, JSONDecodeError handled in parser functions _parse_yaml_file and _parse_json_file respectively and transformed into AirflowFileParseException later. As for FileNotFound and other OSErrors they are handled directly in variables_import function.

Yep, but fail early is still better

Lee-W replied:
> Happy to tackle the description functionality improvements if you're interested and can point me in the right direction.

Not that familiar with this piece, but I would try to see how I can help when I have more bandwidth. Thanks!

evgenii-prusov replied:
> Yep, but fail early is still better

I'm not sure I've got you point. Could you please clarify what exactly do you mean?

Current error handling strategy (as I understand it) is to catch parsing errors from 3 low-level parser functions `_parse_<env|json|yaml>_file`, repackage and return them as lists of `FileSyntaxError` to the caller `_parse_secret_file` and let the caller raise the `AirflowFileParseException` with all collected errors, which let the users see all the issues in their configuration files at once rather than making them fix one by one.

Lee-W replied:
Got it. I just took a deeper look. I just noticed all the cases you mentioned are handled above. Do we still need this `except Exception as e:` block?

evgenii-prusov replied:
As we work with user files, I think it's still useful for catching these exceptions: 
- PermissionError (file access issues)
- OSError (other I/O errors beyond FileNotFoundError)
- MemoryError (for very large files)
- UnicodeDecodeError (encoding issues)

I can see that pattern in connection_command too.

Lee-W replied:
I still like catching explicit exceptions more, but if this pattern is used elsewhere. I'm also ok with it

potiuk replied:
I think it's pretty common pattern in CLI tools. I personally do not like to have ugly (nor even nice) stack trace as result of unhandled exception in CLI so try/except Exception allows to have a nice one line exception message - ideally in 🟥 ....

Lee-W commented on airflow-core/src/airflow/secrets/local_filesystem.py:
```diff
@@ -233,10 +233,23 @@ def load_variables(file_path: str) -> dict[str, str]:
+        if is_env_file and isinstance(secret_values, list):
+            # Only ENV files: Check for duplicate keys
+            if len(secret_values) > 1:
+                raise AirflowException(
```
Let's create a custom exception for it instead of using `AirflowException`

evgenii-prusov committed:
feat(secrets): Improve local filesystem secrets backend error handling and validation

- Add AirflowUnsupportedFileTypeException and VariableNotUnique exceptions
- Use pathlib.Path for file extension parsing instead of string splitting
- Remove manual file existence checks, let Python handle FileNotFoundError
- Fix env file parsing to validate both key and value presence
- Update tests to expect FileNotFoundError for missing files

evgenii-prusov committed:
refactor(cli): Align variable import with LocalFilesystemBackend

- Replace manual JSON parsing with load_variables() function from LocalFilesystemBackend
- Add comprehensive error handling for unsupported file types, parse errors, and duplicate variables
- Update test expectations to match new error messages
- Remove test case for 'list' variable as load_variables treats list values as logical errors
- Update test file extensions to .json for clarity in isolation tests

BREAKING CHANGE: Variable import with description feature is no longer available
after this change due to alignment with LocalFilesystemBackend behavior.

Lee-W commented on airflow-core/src/airflow/cli/commands/variable_command.py:
```diff
@@ -81,11 +86,20 @@ def variables_import(args, session):
+    except AirflowFileParseException as e:
+        raise SystemExit(str(e))
+    except VariableNotUnique as e:
+        raise SystemExit(str(e))
```
```suggestion
    except (AirflowUnsupportedFileTypeException, AirflowFileParseException, VariableNotUnique) as e:
        raise SystemExit(str(e))
```

Lee-W replied:
Then I guess this is the last one we need to handle.

evgenii-prusov committed:
Update airflow-core/src/airflow/cli/commands/variable_command.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>
